### ISSUE TITLE
Dis-tract: Distributed Tract

### ISSUE THREAD

**Issue by czoli1976** [sonos/tract#2465]
@kali I have been toying with the idea of adding plumbing to TRACT to be used as a distributed inference engine across heterogeneous machines with even different backends.

WDYT?

**Comment by czoli1976** [comment]
I spent some time prototyping this end-to-end. Here's what I found, plus a proposed split of responsibilities.

**The distributed-systems layer probably shouldn't live in tract.** Discovery, topology, memory-weighted layer assignment, transport, the API/tokenizer/download loop — that's what a framework like EXO already provides (though heads-up: EXO has become quite MLX/Apple-centric, and its engine contract is a fat `warmup`/`submit`/`step`/`serve_prefill`, so a tract engine there is real work, not a thin shim). tract's job is the part underneath: load a shard of layers, run it on whatever backend that machine has, and hold that shard's KV.

**What tract already makes easy.** Heterogeneous backends essentially fall out of `runtime_for_name("cpu"|"metal"|"cuda").prepare(model)` — a sub-model is backend-neutral NNEF until a worker prepares it, and `run()` takes/returns host tensors (DeviceSync is auto-inserted), so cut-point activations are wire-ready for any backend. KV stays resident inside each worker; only the residual activation crosses between stages.

**The one piece that's genuinely tract's and was missing: a validated way to slice a `TypedModel` into standalone sub-models.** All the primitives exist (`set_input_outlets` / `select_output_outlets` / `translate_model` / `compact`), but there's no validation — a cut whose declared inputs don't dominate its outputs silently yields a wrong sub-model. I prototyped a small addition:

```rust
pub fn extract_subgraph(model: &TypedModel, inputs: &[OutletId], outputs: &[OutletId]) -> TractResult<TypedModel>
pub fn split_pipeline(model: &TypedModel, boundaries: &[Vec<OutletId>]) -> TractResult<Vec<TypedModel>>
```

It's thin (just the primitives above) plus one guard: walk the outputs' dependency cone and error if it reaches an undeclared `Source` — which catches the attention-mask / positional-encoding / external-KV trap, and also a skip connection that spans a cut. No new ops. Statefulness isn't a guard: cutting a stateful op's *tensor* output is fine — the op and its cache stay wholly in one stage.

**Proof it works.** I split a real 16-layer OpenELM at a mid-layer residual across two processes (stage 0 on CPU, stage 1 on Metal), with each worker looping its own KV locally, and get token-for-token parity vs the single-machine run (prefill logits cosine 1.0, argmax identical through greedy decode). Each worker holds only its half of the weights, so no single machine holds the whole model.

A few things I'd want your read on before shaping this into anything:

1. Does a subgraph-extraction primitive belong in `tract-core` (`model` module), or is it better as a `tract-libcli` / example utility?
2. Free functions vs methods on `TypedModel`?
3. Typed error variants vs plain anyhow context?

Everything above the cut line (partitioning strategy, transport, worker orchestration) I've kept in a separate crate for now — glad to go whichever way you think fits what should live in tract.

**Comment by czoli1976** [comment]
<img width="1128" height="481" alt="Image" src="https://github.com/user-attachments/assets/cef48f4e-6f97-460d-abd7-e1723fc6d727" />

**Comment by kali** [comment]
We just have to do it, the name is too good to waste.

So we need a way to split models. I think there are two sub-problems. One is deciding where to split, the second is implementing the split proper.

Then we have an interface question: public API or internal ones.
And finally the placement thing. Is this comparable to the causal_llm thing ? or a monster bigger than tract itself that lives outside and just consumes tract ?
 
My inclination, unless it's already huge: like causal_llm. In tract repo, using the public API. That way it's relatively easy to spin if off to a separate repository. And it puts the kind of pressure we need on the public API to figure out what it lacks.

For the proper split implementation, I /think/ the public API exposes what we need already in the form of the select input and select output transforms. Note that it operates on node names (which can be very finicky because of the renaming happening at declutter time) and not outletid. The public API does not offer any graph walking routines, maybe that's what coming next.

If we want to lower the split below the API, for stronger consistency and ability to walk the graph nodes, I think Split should be a Transform: not a TypedModel method, not a freestanding function, but a registered Transform that can be discovered and called through the public API.  The Split could generate a model containing multiple independent models, with input/output convention names, that would need to be isolated with select input and select outputs. Or it could be SplitAndExtract, that would be called multiple times with a shard index to extract multiple subgraphs.

The main reason why I'm so keen on pushing Transform is... it's basically a simple method call serialisation protocol. So the python and C get them for free. Transform is meant to be an extension point to allow plugging stuff below the API that is accessible through the API. We already have first-class-citizen Transforms (with a type safe Configuration client side) but also "experimental" transforms relying on the serialized transform configuration.

As for error reporting, my stance is, start with Anyhow and Strings, replace the String root cause by a typed error only when it proves it is useful. Very few errors are recoverable anyway, typing errors is akin to premature optimisation imho.

WDYT ?

**Comment by czoli1976** [comment]
ok, let me toy with it a bit more then I can share a prototype, yeah the name is soo cool, I could not resist

**Comment by czoli1976** [comment]
@kali  Prototyped it end-to-end (pipeline/layer-split, 2+ workers, heterogeneous cpu+metal, real instruct models chatting through the cluster). Since you wanted this to "put pressure on the public API to figure out what it lacks" — here's exactly what it lacked, grouped the way you framed it: the split itself, then everything else building it surfaced. I probably got a few things wrong so here this recap to discuss (we can do it over the PR, if you prefer)

## 1. Placement & form — agree

causal_llm-style, in-repo, on the public API, easy to spin off later. Dis-tract already builds this way (an additive `tract-distributed` crate). And I'm sold on **Split as a registered Transform** — the "it's a method-call serialization protocol, so Python and C get it for free" argument is the decider; a freestanding function throws that away.

My prototype currently dogfoods two free functions instead:

```rust
pub fn extract_subgraph(model: &TypedModel, inputs: &[OutletId], outputs: &[OutletId]) -> TractResult<TypedModel>
pub fn split_pipeline(model: &TypedModel, boundaries: &[Vec<OutletId>]) -> TractResult<Vec<TypedModel>>
```

= the existing primitives (`set_input_outlets` / `select_output_outlets` / `translate_model` / `compact`) plus **one boundary-validation guard**: walk the outputs' dependency cone and error if it reaches an undeclared Source (a cut whose declared inputs don't dominate its outputs otherwise silently yields a wrong sub-model). Happy to refold this behind a registered `Split` / `SplitAndExtract(shard_idx)` Transform. Of your two shapes, `SplitAndExtract(idx)` fit the distributed case more cleanly — each worker asks for exactly its shard, no multi-model-then-isolate dance — but the isolate-with-select-in/out route needs the naming convention nailed down first (see §2).

## 2. You're right about node-names-vs-OutletId — it's the #1 real pain

The most expensive bug in the whole prototype was exactly this. Deciding *where* to split means knowing which layer a node belongs to, and with only node names to go on I had to parse them — `model_model__15_selfAttn_keyStates_1` parsed to layer **1**, not 15, so layer-15 nodes got labeled stage 0, over-collected across the cut, and silently broke the working openelm split. It took many rounds to find. I ended up with a bespoke `node_layer()` heuristic (`__{N}_` first, then `layers_{N}_`, then cache last-segment, else None) that's pure guesswork against the current exporter's renaming.

This is the "no graph-walking routines yet" gap you flagged, and it's the thing I'd most want from the public API:

- **(a) Graph-walking / dominance utilities** — cone-of-influence, dominators, "does this input set dominate this output set" — so partitioning and the validation guard above aren't reimplemented per consumer.
- **(b) Block/layer boundary metadata** — a queryable notion of "which transformer block owns this node," so deciding *where* to split doesn't mean reverse-engineering layer membership from names that declutter is free to rewrite. This is what makes the OutletId-based path robust instead of the name-based one being finicky.

## 3. Other public-API gaps found building the rest end-to-end

Not split-specific, but each one was a wall:

1. **No canonical "prepare" entry point.** The `transformers_detect_all → unfold-kv-cache → declutter/optimize` sequence only lives inside the causal_llm *example*, not as a callable. My shard loader skipped `declutter` because that wasn't obvious, and paid **2.8× on Metal** (MetalTransform's rewrite rules only match decluttered graphs — 73 `MultiBroadcastTo` stranded on host, 5364 nodes vs 1690). A shard *must* be prepared identically to a whole model; a supported public `prepare` fn would make that guaranteed instead of copy-the-example.

2. **KV-cache endpoints aren't introspectable.** I pair them by string-stripping `in_cache_*_N` / `out_cache_*_N` and infer the cache axis by convention. A structured API — enumerate the in/out pairs, their layer index, and the cache/seq axis — would replace the most exporter-fragile part of a worker.

3. **Serialization isn't round-trip-safe for transformed models.** Shipping a shard via `model_to_bytes` fails *after* `transformers_detect_all` on Qwen: a mixed-dtype EinSum trips `einsum-to-prefix-matmul … require operating_dt F16`, blocking even single-node. I had to sidestep it entirely (workers load the full NNEF and extract their range locally). A faithful re-serialize of an already-transformed model — or a first-class "serialize a sub-model + reload it exactly" — would unblock actually *shipping* shards.

4. **No lazy / partial weight loading.** Because a worker loads the full model then extracts, load-time memory doesn't scale with shard count — which defeats the point (fitting a too-big model, not just running one). Lazy NNEF weight loading, so a worker materializes only its shard's weights, is the missing piece for the memory story.

5. **Symbolic-shape (`TDim`) values across a boundary.** Cut activations with `datum_type == TDim` can't serialize (`TDim vs TDim`), so I filter them out of the forwarded set and recompute per stage. Worth a documented intended path for moving/reconstructing symbolic-shape values across a serialized cut.

6. **Backend selection needs `extern crate`.** The runtime contract itself is great — `runtime_for_name("cpu"|"metal"|"cuda").prepare(model).spawn()`, host tensors in/out with DeviceSync auto-inserted, so cut activations are wire-ready for any backend. The only wart: a worker must force-link backends via cfg-gated `extern crate tract_metal/tract_cuda`. Self-registration so they're selectable by name would kill the hack.

## Suggested first step

Land `Split` / `SplitAndExtract` as a registered Transform (§1) on top of a couple of graph-walking routines (§2a) — that's the smallest thing that turns the name-based finickiness into an OutletId-based, discoverable, Python/C-reachable operation, and it's the piece Dis-tract is already dogfooging. The §3 items I can file as separate focused issues if you'd rather track them independently.

**Comment by czoli1976** [comment]
@kali , no urgency here, just making you aware

**Comment by kali** [comment]
I'm aware :) I defintely want to have a look, but I'm trying to close the performance regressions bits before opening this new can of worms.

**Comment by czoli1976** [comment]
The protein of the future
